### PR TITLE
Obey alignment in matrices

### LINKED ISSUE #460: How does alignment actually work?
The documentation simply says

https://typst.app/docs/reference/math/

"Formulas can also contain line breaks. Each line can contain one or multiple alignment points (&) which are then aligned."

Coming from LaTeX the following
```
\begin{alignat*}{2}
        a &=b &\quad c&=d\\
        e&=f & g&=h
    \end{alignat*}
```
is displayed as

image

Therefore I expected

```
$
a &=b & quad c&=d \
e &=f & g&=h
$
```
to look similar. However, it's displayed as

image

Am I missing something?

Is it not based on alternating right and left aligned columns, like in LaTeX? Or is this a bug?

### PR DESCRIPTION
**Pull request by SUPERCILEX**
Makes progress on #460

### CODE REVIEW

**Comment by laurmaedje** [comment]
Wait, how is this related to #460, which talks about general equation alignment whereas this changing the behaviour of matrics?

**Comment by Enivex** [comment]
Arbitrary alignment in matrices is something I've not seen before. I think most people would be happy with just being able to globally specify the alignment of the columns. (Or possibly the alignment of each column.)

**Comment by SUPERCILEX** [comment]
> Wait, how is this related to #460, which talks about general equation alignment whereas this changing the behaviour of matrics? Edit: Okay, that issue talks about both things.

Oops, updated the PR description to not close that issue. But yeah, it handles half of it by flip flopping matrices.

> Arbitrary alignment in matrices is something I've not seen before

This PR isn't doing anything with that, it's just part of the test suite so we're aware of the current behavior.

**Comment by SUPERCILEX** [comment]
Ok, fixed the merge conflicts.

**Comment by laurmaedje** [comment]
For equation alignment with `&`, I think alternating makes sense, but for matrices it seems strange. They should just have an `align` parameter, e.g. `mat(-1, 1; 1, -1; align: #left)`. In particular, LaTeX also doesn't do this. I tried with this example:
```tex
\begin{align}
\begin{pmatrix}
  -1 &  1 &  1 \\
   1 & -1 &  1 \\
   1 &  1 & -1 \\
\end{pmatrix}
\end{align}
```
It defaults to center alignment like Typst.

**Comment by SUPERCILEX** [comment]
Ah shoot, I misread the issue and didn't see the `[l]` in the latex code. I've updated this PR to keep center alignment if there are no `&` in the matrix and otherwise use the `&`s. This is consistent with how equations work (see the test). 

Regarding an alignment parameter, I don't think it's necessary with tabular alignment because people can use `&` in the first row of their matrix and get precisely the behavior they want for each column.

**Comment by Enivex** [comment]
> Regarding an alignment parameter, I don't think it's necessary with tabular alignment because people can use `&` in the first row of their matrix and get precisely the behavior they want for each column.

Not if you intend alignment to work as in latex. You will be able to do center (no &), but you can only achieve flush-right by adding & only in the first row. The implicit &s will always be added at the end.

**Comment by SUPERCILEX** [comment]
> Honestly I think &s should be removed from matrices entirely

I'm not sure I agree with this as `&` can provide sub-alignment within each column.

> You cannot achieve flush left without adding &s to every cell in the column.

Crap, you're right. Then I agree that we should have an align param, but I don't think that needs to happen in this PR.

File: tests/typ/math/alignment.typ [hunk 1]
```diff
@@ -12,17 +12,115 @@ $
...
+---
+// Test alternating alignment in a vector.
+#set page(width: 250pt)
+$
+vec(
+"abc" & "abc abc abc" & "abc abc",
+"abc abc" & "abc abc" & "abc",
+"abc abc abc" & "abc" & "abc abc abc",
+)
+$
```
Comment (laurmaedje) [inline tests/typ/math/alignment.typ:49]:
```suggestion
$ vec(
"abc" & "abc abc abc" & "abc abc",
"abc abc" & "abc abc" & "abc",
"abc abc abc" & "abc" & "abc abc abc",
) $
```

File: tests/typ/math/alignment.typ [hunk 2]
```diff
@@ -12,17 +12,115 @@ $
...
+mat(&a+b&,c;&d&, e)
+$
+
+$
+mat(&&&a+b,c;&&&d, e)
+$
+
+$
+mat(.&a+b&.,c;.....&d&....., e)
+$
```
Comment (laurmaedje) [inline tests/typ/math/alignment.typ:116]:
```suggestion
$ mat(&a+b, c; &d, e) $
$ mat(&a+b&, c; &d&, e) $
$ mat(&&&a+b, c; &&&d, e) $
$ mat(.&a+b&., c; .....&d&....., e) $
```

**Review by laurmaedje — COMMENTED** [review]
Okay, now this is simpy a bug fix, since I intended for alignment points to work, but then apparently broke that by centering everything. I have one request before merging though: The tests use a bit different and more compact math style, which I illustrated with two suggestions. Could you apply that across the file? I also feel like the tests could be made a bit more compact and understandable if they wouldn't use "abc" everywhere.

**Pull request merged**